runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,449 @@
+                     @Nullable final Timestamp timestamp) {
+        try {
+            txBegin(timestamp);
+            return getTable(tableName).count();
+        } finally {
+            txEnd();
+        }
+    }
+
+    @Nonnull
+    private <K extends Message, V extends Message, M extends Message>
+    Collection<CorfuRecord<V, M>> scanAndFilter(@Nonnull final String tableName,
+                                                @Nullable Timestamp timestamp,
+                                                @Nonnull final Predicate<CorfuRecord<V, M>> p) {
+        try {
+            txBegin(timestamp);
+            return ((Table<K, V, M>) getTable(tableName)).scanAndFilter(p);
+        } finally {
+            txEnd();
+        }
+    }
+
+    /**
+     * Query by a secondary index.
+     *
+     * @param tableName Table name.
+     * @param indexName Index name. In case of protobuf-defined secondary index it is the field name.
+     * @param indexKey  Key to query.
+     * @param <K>       Type of Key.
+     * @param <V>       Type of Value.
+     * @param <I>       Type of index/secondary key.
+     * @return Result of the query.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message, I extends Comparable<I>>
+    QueryResult<Entry<K, V>> getByIndex(@Nonnull final String tableName,
+                                        @Nonnull final String indexName,
+                                        @Nonnull final I indexKey) {
+        return new QueryResult<>(((Table<K, V, M>) getTable(tableName)).getByIndex(indexName, indexKey));
+    }
+
+    private <V extends Message, M extends Message, R>
+    Collection<R> initializeResultCollection(QueryOptions<V, M, R> queryOptions) {
+        if (!queryOptions.isDistinct()) {
+            return new ArrayList<>();
+        }
+        if (queryOptions.getComparator() != null) {
+            return new TreeSet<>(queryOptions.getComparator());
+        }
+        return new HashSet<>();
+    }
+
+    private <V extends Message, M extends Message, R>
+    Collection<R> transform(Collection<CorfuRecord<V, M>> queryResult,
+                            Collection<R> resultCollection,
+                            Function<CorfuRecord<V, M>, R> projection) {
+        return queryResult.stream()
+                .map(v -> Optional.ofNullable(projection)
+                        .map(function -> function.apply(v))
+                        .orElse((R) v))
+                .collect(Collectors.toCollection(() -> resultCollection));
+    }
+
+    /**
+     * Execute a scan and filter query.
+     *
+     * @param tableName Table name.
+     * @param query     Predicate to filter the values.
+     * @param <V>       Type of Value.
+     * @param <R>       Type of returned projected values.
+     * @return Result of the query.
+     */
+    @Nonnull
+    public <V extends Message, M extends Message, R>
+    QueryResult<R> exectuteQuery(@Nonnull final String tableName,
```

Comment:
If executeQuery is only for scanAndFilter, why not just name it scanAndFilter to be consistent with other APIs like get() and getByIndex()?  Naming it executeQuery  sounds like JDBC way of  doing query, which seems weird in this context since it's not consistent with other APIs.

Reply:
You are right but this was an explicit ask to be consistent with the way the Query module is in use by our current consumers in different verticals.